src/managers/health.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling service health."""
+
+import logging
+import os
+
+import requests
+from requests.exceptions import ConnectionError, HTTPError
+
+from core.cluster import SUBSTRATES, ClusterState
+from core.workload import WorkloadBase
+from exceptions import OSDAPIError
+from literals import (
+    HEALTH_OPENSEARCH_STATUS_URL,
+    MSG_STATUS_DB_DOWN,
+    MSG_STATUS_DB_MISSING,
+    MSG_STATUS_ERROR,
+    MSG_STATUS_UNAVAIL,
+    MSG_STATUS_UNHEALTHY,
+    MSG_STATUS_UNKNOWN,
+    MSG_STATUS_WORKLOAD_DOWN,
+    OPENSEARCH_CA_FILE,
+)
+from managers.api import APIManager
+
+logger = logging.getLogger(__name__)
+
+
+class HealthManager:
+    """Manager for handling Kafka machine health."""
+
+    def __init__(
+        self,
+        state: ClusterState,
+        workload: WorkloadBase,
+        substrate: SUBSTRATES,
+    ):
+        self.state = state
+        self.workload = workload
+        self.substrate = substrate
+        self.api_manager = APIManager(state, workload, substrate)
+
+    def status_ok(self) -> tuple[bool, str]:
+        """Health status"""
+        try:
+            status_data = self.api_manager.service_status()
+        except HTTPError as err:
+            if err.response.status_code == 503:
+                return False, MSG_STATUS_UNAVAIL
+        except (ConnectionError, OSDAPIError):
+            return False, MSG_STATUS_UNAVAIL
+
+        if status_data["status"]["overall"]["state"] == "green":
+            return True, ""
+        elif status_data["status"]["overall"]["state"] == "yellow":
+            return True, MSG_STATUS_UNHEALTHY
+        elif status_data["status"]["overall"]["state"] != "green":
+            return False, MSG_STATUS_ERROR
+        return True, MSG_STATUS_UNKNOWN
+
+    def opensearch_ok(self) -> tuple[bool, str]:
+        """Verify if associated Opensearch service is up and running."""
+
+        if not self.state.opensearch_server or not self.state.opensearch_server.tls_ca:
+            return False, MSG_STATUS_DB_MISSING
+
+        if not os.path.isfile(OPENSEARCH_CA_FILE):
+            with open(OPENSEARCH_CA_FILE, "w") as text_file:
+                text_file.write(self.state.opensearch_server.tls_ca)
```

Comment:
Please be aware that the CA of Opensearch might change (CA rotation), that means the file you have locally could be out of date. This would then lead to a verification error on the request.

I would recommend to get and write the file each time, not just if it's missing.

Reply:
Thank U so much for keeping this one in mind.

Indeed we need to check it each time fresh. Also I should use the file write method from `workload`. Thanks much!!!

Reply:
All right. In fact, I just noticed that I'm already caching the Opensearch CA file from when the relation is joined.

And that one is updated everytime the Opensearch relation (data) has changed.

So we shoudl be good :-)